### ISSUE TITLE
Home -> Navigation -> Problem: Inaccessible Child Controls 

### ISSUE THREAD

**Issue by johnhbenetech** [benetech/MathShare#762]
Depending on the user’s current context (e.g. editing a problem set), several controls can appear inside the problem button.  These controls are: 
 
• an image button, to open the problem’s scratchpad; 
• an Edit button, to open a modal dialog allowing the problem’s metadata to be edited; 
• a remove button, which prompts the user to confirm that they wish to delete a problem;
• a control which in the source code is referred to as the “plus button”, the purpose of which is unknown. 
 
All of the above controls are inaccessible to a screen reader user, because they are marked up as children of the < button > representing the problem in the list.  In addition, they aren’t rendered as buttons themselves, nor do they have accessible labels.  They are, however, forced to appear in the tab order using a tabindex attribute.  Finally (although not directly related to accessibility), the problem’s < button > does contain several < div> elements as children, which is not valid HTML. 
 
The end result is that a screen reader user will be unaware of the presence of these controls unless they tab through the list, at which point they will encounter several unlabelled, silent areas which cause no speech.  At that point, they will be aware that some sort of element is present, but have no way of gauging what it is, what it’s for or how to use it.

Aha! Link: https://diagramlabs.aha.io/features/B-131

**Comment by johnhbenetech** [comment]
Suggested Solution 
 
Visually, these controls are implemented as clickable icons, and a design decision has presumably been made that they don’t clutter the space.  The issue is that if they are implemented as accessible buttons with textual labels, they will make the list crowded and confusing to navigate. 
 
Therefore, in addition to improving their accessibility, we would advise wrapping these buttons inside a togglable sub-menu, following a similar implementation pattern to that recommended for the problem set controls.  In short: 

* Each control should be marked up as a button (using a native < button> element) if it causes a same-page action, or a hyperlink (using a native < a> element) if it causes navigation to a different route. 
* Each button must have an accessible textual label, preferably implemented as visible, on-screen text to allow easy activation by users of speech input technology.  Otherwise, the label can be provided as the value of an aria-label attribute. 
* Each label should be integrated into the localization system via an entry in strings.js. 
* The buttons should be contained inside an unordered list, with each one marked up inside a < li> list item. 
* The opening < ul> tag of the sub-menu should have an aria-labelledby attribute, referencing the ID of the corresponding toggle button. 
* The toggle button, used to collapse and expand the sub-menu, should: 
o also be marked up as a < button>; 
o have an accessible textual label which incorporates the number of the problem, for example: “Options for Problem 1”; 
o include an aria-expanded attribute, the value of which should initially be set to “false”, updated to “true” when the sub-menu is being displayed and back to “false” when collapsed again; 
o not include an aria-haspopup attribute; and 
o have a unique ID which can be applied as the value of an aria-labelledby attribute on the sub-menu container’s opening < ul> tag to provide an accessible name. 

* The toggle button should only be shown when the sub-menu will be non-empty based on context.  For instance, if none of these controls should be available while viewing a problem set, the toggle button shouldn’t be available either. 
 
Note that the label text of the “Options for …” button should not include the name of the problem, as this would make for an overly long label which was difficult to parse.  The approach outlined above is shown in the proposed code sample.

**Comment by johnhbenetech** [comment]
Proposed Code Sample 
 
File: src/components/Home/components/Navigation/Problem/index.js

```
const NavItem = withRouter(({ history }) => (  <div id={`problem-${this.props.number + 1}`} className={…}>   <button className={…} onClick={…} onKeyPress={…}>    <span className={…}>     {wrappedAnnotation}     … equation span …     {speechForMath}    </span>   </button>   <div className="dropdown">    <button id={`dropdownMenuButton-${this.props.number + 1}`} aria-expanded="false" …>     Options for Problem {this.props.number + 1}    </button>    <ul className={problemStyle.btnContainer} arialabelledby={`dropdownMenuButton-${this.props.number + 1}`}>     {imgButton}     {editButton}     {removeButton}     {plusButton}    </ul>    {this.props.problem && this.props.problem.scratchpad ? image : null}   </div>  </div> )); 
 
const imgButton = (this.props.problem && this.props.problem.scratchpad)  ? (   <a href="#" onClick={this.onImgClick} …>    <span>{Locales.strings.problem_image}</span>    <span>&nbsp;({Locales.strings.opens_in_new_window})</span>    <FontAwesome className={classNames(problemStyle.imgIcon, 'fa-2x')} name="image" />   </a>  )  : null; 
const plusButton = this.props.addNew  ? (   <button …>    {Locales.strings.plus_problem} /* (or another key name which makes more sense) */    <FontAwesome className={classNames(problemStyle.plusIcon, 'fa-2x',)} name="plus-circle" />   </button>  )  : null; 
 
const editButton = this.props.showRemove  ? (   <button onClick={this.onEditClick} …>    {Locales.strings.edit_problem}    <FontAwesome className={classNames(problemStyle.editIcon, 'fa-2x')} name="edit" />   </button>  )  : null; 
 
const removeButton = this.props.showRemove  ? (   <button onClick={this.onTrashClick} …>    {Locales.strings.remove_problem}    <FontAwesome className={classNames(problemStyle.trashIcon, 'fa-2x')} name="trash" />   </button>  )  : null; 
```

**Comment by rupeshparab** [comment]
@johnhbenetech 
Some issues mentioned at https://github.com/benetech/MathShare/issues/740 have been fixed in this
I have fixed the issues which don't have a significant UI change
But one of the change required is to move all buttons (add, edit, trash, scratchpad) to a dropdown
Please let me know if it is ok

**Comment by johnhbenetech** [comment]
@alexrcabral @benetechMatt 

We need a new design approval/recommendation to remediate these issues.

Currently a problem "box" in a problem set "grid" can have several sub element controls depending on the context.

* edit problem (on edit set)
* delete problem (on edit set)
* add problem (on edit set)
* view sketchpad image (on both edit and view set)

Please review the above recommendation and let us know what you want. Keep in mind that in the case of the 'edit problem set' page there may be as many as 4 sub elements. In the case of the 'view problem set' page there will only ever be zero or the sketchpad sub element.

**Comment by alexrcabral** [comment]
Add problem should not be a separate button, since the box itself functions as a button leading to adding a new problem. Same with edit problem. 

Delete problem is the only one that should remain. Rather than having 4 sub elements in a list, there should only be one sub element to "delete". 

Could you elaborate on "view sketchpad image"? Is this just the function of the box as a whole, which when clicked on functions as a link to edit the problem? Assuming that's the case:

A sub-menu isn't necessary. The plus button and edit buttons are redundant and should be removed, while the problem boxes retain function & read out their function when in focus for screen reader users.

**Comment by johnhbenetech** [comment]
Currently 'edit' and clicking on the box do two separate things (or at least bring up two separate modals). Edit allows you to change the actual prompt. Clicking the box only allows you to view the problem as a student. So we probably need some changes there if we want to avoid a menu item - we would also need to make clear that's what clicking the box does.

Re: sketchpad. If you attach a sketchpad drawing to the prompt, an icon for the sketch appears in the box (on both edit and student view)

Edit view icon:
![image](https://user-images.githubusercontent.com/12927553/65985215-00dff900-e447-11e9-9d1f-2519a5a8408f.png)

Student view:
![image](https://user-images.githubusercontent.com/12927553/65985260-11906f00-e447-11e9-84e3-faeb49ce2e02.png)

**Comment by johnhbenetech** [comment]
@benetechMatt please advise on path we want to take here

**Comment by benetechMatt** [comment]
- Clicking on the box will allow teachers to edit the problem. No longer will go the problem solution page.  Student preview meets the need
- Remove edit problem (no longer needed).
- for the image, remove the icon and have a thumbnail of the image appear instead. same padding as text and should not expand the box height.
- move delete icon to upper right. functions the same
- add problem will only be available as it's own button, no change.

![image](https://user-images.githubusercontent.com/49413678/66521004-e8f41f00-eaaf-11e9-999d-e152c3ae76ca.png)

**Comment by johnhbenetech** [comment]
@benetechmatt what should image Alt text for sketch be?

**Comment by benetechMatt** [comment]
Diagram for problem.

Makes me think we should add an alt text field for drawings.  Will create a new issue for that

________________________________

**Comment by johnhbenetech** [comment]
@benetechMatt Per the recommendation, it would be best if the delete icon had a corresponding label like 'Delete'. Does that sound good? If so are you still fine leaving it in the upper right corner?

**Comment by benetechMatt** [comment]
@alexrcabral and I discussed this and decided that it would be more consistent if we adopt a three dot ellipses for options.  This will also allow us to add additional options in the future.

Below is an image showing the plan.  There will only be two options at this time: edit, and delete.
![two problem cards.  Card on the left has the three dot ellipses in the lower right.  Card on the right is the same but with the ellipses clicked and displaying the card options menu.  options are edit, add step (coming later), and delete. The three items have icons.](https://user-images.githubusercontent.com/49413678/67712455-ffe0af80-f991-11e9-9321-810c396fb706.png)

Clicking on the problem lets teachers view the problem.  Edit brings up the current edit modal for problems.  delete gives users the same modal asking if they want to delete the problem.


### Diagrams:
@jscholes currently the images open in a new window.  Alex and I need to think more about how we want to use/handle images.

**Comment by rupeshparab** [comment]
I have moved the edit and delete buttons inside a dropdown and haven't made any changes to the scratchpad image as it is currently in discussion.

I have added the following accessibility changes while adding the dropdown:
On dismissal of the edit modal, the focus would return to the dropdown of the problem being edited. In case of the delete confirmation modal it would return to the next problem moved to the same position or the last problem in case the last problem is deleted.
Also, I have added this text for the dropdown: 'More Options for {title}', here title is the problem title.

**Comment by benetechMatt** [comment]
I like @jscholes suggestion of "view problem diagram" as the button label action. 
created issue #855 for future work on adding image descriptions

**Comment by jscholes** [comment]
I'm happy to say that the accessibility of the controls has been addressed.  But overall I'm not clear where we stand on problem sketches (or diagrams).  Are there any problems in any of the defined sets with diagrams populated, and if so, where are they shown in relation to these dropdown options?

**Comment by jscholes** [comment]
CC @benetechMatt, @johnhbenetech and @alexrcabral

**Comment by johnhbenetech** [comment]
@jscholes There are some technical bugs introduced in QA with a recent change. Once those are remediated I'll create an example problem set with sketches so you can review how it is afforded. Thanks for your patience!

**Comment by johnhbenetech** [comment]
Hey @jscholes, sorry for the delay on this. The following is an example where problem 1 has a sketch and problem 2 does not.


[Link to example with sketch](https://mathshare-qa.diagramcenter.org/#/app/problemSet/view/EMR2VMO3HS7RS)

**Comment by jscholes** [comment]
@johnhbenetech Thanks for that sample set.  I see that when editing it, the "View Sketch" button isn't positioned inside the options dropdown for the problem.  Is that the design we're aiming for?  Of course, having it inside a dropdown on the solve view wouldn't make sense, because it would be the only option.

Additionally, the label of "View Sketch" doesn't indicate to which problem it relates.  As a minimum, the button should have an accessible label of "View Sketch for title" (same as the dropdown buttons), e.g. "View Sketch for Example 1".

CC @rupeshparab

**Comment by johnhbenetech** [comment]
@benetechMatt We need a final decision on this to wrap this up.

The options seem to be:

1. Keep the current approach, where the button/icon stays in the problem tile, but just change the alt/label to "View Sketch for Problem 1" etc

2. Move the 'View Sketch for Problem 1" into the three dot menu on both the edit/solve view. It would be the only option in the menu in the solve view.

Please let us know your preference

**Comment by jscholes** [comment]
To be clear, these options are technically correct, but we would struggle to sign off on `#2` unless additional options are planned in the immediate future for the dropdown on the solve view.  Adding an extra click just to view a single-option menu has no net benefit.

**Comment by johnhbenetech** [comment]
We will go with option 1 above due to lack of communication from product team

**Comment by alexrcabral** [comment]
Completed and released as part of v1.1.0
